### PR TITLE
ORC: Supported nested identity partition data 

### LINKED ISSUE #897: ORC: Support partition values from a constant map
Identity partition values should be added to materialized records using values from a file's partition data. The initial implementation for Spark used a `JoinedRow` to join the partition values to each row read from a format, but this had a few problems:

* Only top-level fields could be set this way, not nested fields
* Values were not added in place and would require a projection in Spark
* No support for directly reading tables with generics

Avro and Parquet are moving to implementations that pass a map from field ID to a value when building the reader, so the constant can be added at the right place in the read schema, and so that the implementation can be shared across in-memory representations. ORC should also add support for passing partition values as a map.

### PR DESCRIPTION
**Pull request by rdsr**
Fixes #897

The  following changes are made
1. Added a OrcTypeWithSchemaVisitor
2. Added support for nested identity partition similar to #896 
3. Refactored a lot of data reader functions which can now be shared across Spark and Iceberg Generics for ORC
4. SparkOrcReader is simplified in the process. I've moved away from  Spark `UnsafeRow` and used GenericInternalRow instead similar to Spark Avro and Parquet readers.

### CODE REVIEW

File: orc/src/main/java/org/apache/iceberg/orc/OrcValReader.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.orc;
+
+import org.apache.orc.storage.ql.exec.vector.ColumnVector;
+
+
+public interface OrcValReader<T> {
```
Comment (rdsr) [inline orc/src/main/java/org/apache/iceberg/orc/OrcValReader.java:25]: I want to rename this to OrcValueReader, but an interface with that name already exists. I though I'd rename the other interface OrcRowReader instead. If folks are OK doing this as part of this RB, let me know.

File: orc/src/main/java/org/apache/iceberg/orc/OrcValueReaders.java [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+import org.apache.iceberg.types.Types;
+import org.apache.orc.storage.ql.exec.vector.BytesColumnVector;
+import org.apache.orc.storage.ql.exec.vector.ColumnVector;
+import org.apache.orc.storage.ql.exec.vector.DoubleColumnVector;
+import org.apache.orc.storage.ql.exec.vector.LongColumnVector;
+import org.apache.orc.storage.ql.exec.vector.StructColumnVector;
+import org.apache.orc.storage.ql.exec.vector.VectorizedRowBatch;
+
+
+public class OrcValueReaders {
```
Comment (rdsr) [inline orc/src/main/java/org/apache/iceberg/orc/OrcValueReaders.java:35]: `GenericOrcReader` can share a lot of the code here.

File: spark/src/main/java/org/apache/iceberg/spark/data/SparkOrcValueReaders.java [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+
+    protected StructReader(List<OrcValReader<?>> readers, Types.StructType struct, Map<Integer, ?> idToConstant) {
+      super(readers, struct, idToConstant);
+      this.numFields = readers.size();
+      internalRow = createOrReuse();
+    }
+
+    @Override
+    protected InternalRow createOrReuse() {
+      return new GenericInternalRow(numFields);
```
Comment (rdsr) [inline spark/src/main/java/org/apache/iceberg/spark/data/SparkOrcValueReaders.java:134]: My guess would be this is what is tanking the perf. Spark[Avro|Parquet]Reader seem to reuse this object. Doing the same for ORC results in test failures

**Comment by rdsr** [comment]
cc @edgarRd

File: orc/src/main/java/org/apache/iceberg/orc/OrcSchemaWithTypeVisitor.java [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+  }
+
+  public static <T> T visit(Type iType, TypeDescription schema, OrcSchemaWithTypeVisitor<T> visitor) {
+    switch (schema.getCategory()) {
+      case STRUCT:
+        return visitRecord(iType.asStructType(), schema, visitor);
+
+      case UNION:
+        // We don't have an answer for union types.
+        throw new IllegalArgumentException("Can't handle " + schema);
```
Comment (abti) [inline orc/src/main/java/org/apache/iceberg/orc/OrcSchemaWithTypeVisitor.java:42]: how about using UnsupportedOperationException instead?

**Comment by rdsr** [comment]
I reran the Spark jmh tests which @shardulm94  had written on a fresh checkout of iceberg and against my patch. Here's the output

Fresh checkout of Iceberg
```
Benchmark                                                                          Mode  Cnt   Score   Error  Units
IcebergSourceNestedORCDataReadBenchmark.readFileSourceNonVectorized                  ss    5  11.004 ± 0.330   s/op
IcebergSourceNestedORCDataReadBenchmark.readFileSourceVectorized                     ss    5  10.920 ± 1.424   s/op
IcebergSourceNestedORCDataReadBenchmark.readIceberg                                  ss    5   2.028 ± 0.103   s/op
IcebergSourceNestedORCDataReadBenchmark.readWithProjectionFileSourceNonVectorized    ss    5  10.695 ± 0.203   s/op
IcebergSourceNestedORCDataReadBenchmark.readWithProjectionFileSourceVectorized       ss    5  10.293 ± 0.236   s/op
IcebergSourceNestedORCDataReadBenchmark.readWithProjectionIceberg                    ss    5   1.850 ± 0.211   s/op
```

With the ORC nested partition patch applied
```
Benchmark                                                                          Mode  Cnt   Score   Error  Units
IcebergSourceNestedORCDataReadBenchmark.readFileSourceNonVectorized                  ss    5  13.984 ± 0.440   s/op
IcebergSourceNestedORCDataReadBenchmark.readFileSourceVectorized                     ss    5  10.454 ± 0.438   s/op
IcebergSourceNestedORCDataReadBenchmark.readIceberg                                  ss    5   2.264 ± 0.091   s/op
IcebergSourceNestedORCDataReadBenchmark.readWithProjectionFileSourceNonVectorized    ss    5  10.037 ± 0.241   s/op
IcebergSourceNestedORCDataReadBenchmark.readWithProjectionFileSourceVectorized       ss    5  10.615 ± 0.336   s/op
IcebergSourceNestedORCDataReadBenchmark.readWithProjectionIceberg                    ss    5   1.980 ± 0.167   s/op
```

**Comment by rdsr** [comment]
> I reran the Spark jmh tests which @shardulm94 had written on a fresh checkout of iceberg and against my patch. Here's the output
> 
> Fresh checkout of Iceberg
> 
> ```
> Benchmark                                                                          Mode  Cnt   Score   Error  Units
> IcebergSourceNestedORCDataReadBenchmark.readFileSourceNonVectorized                  ss    5  11.004 ± 0.330   s/op
> IcebergSourceNestedORCDataReadBenchmark.readFileSourceVectorized                     ss    5  10.920 ± 1.424   s/op
> IcebergSourceNestedORCDataReadBenchmark.readIceberg                                  ss    5   2.028 ± 0.103   s/op
> IcebergSourceNestedORCDataReadBenchmark.readWithProjectionFileSourceNonVectorized    ss    5  10.695 ± 0.203   s/op
> IcebergSourceNestedORCDataReadBenchmark.readWithProjectionFileSourceVectorized       ss    5  10.293 ± 0.236   s/op
> IcebergSourceNestedORCDataReadBenchmark.readWithProjectionIceberg                    ss    5   1.850 ± 0.211   s/op
> ```
> 
> With the ORC nested partition patch applied
> 
> ```
> Benchmark                                                                          Mode  Cnt   Score   Error  Units
> IcebergSourceNestedORCDataReadBenchmark.readFileSourceNonVectorized                  ss    5  13.984 ± 0.440   s/op
> IcebergSourceNestedORCDataReadBenchmark.readFileSourceVectorized                     ss    5  10.454 ± 0.438   s/op
> IcebergSourceNestedORCDataReadBenchmark.readIceberg                                  ss    5   2.264 ± 0.091   s/op
> IcebergSourceNestedORCDataReadBenchmark.readWithProjectionFileSourceNonVectorized    ss    5  10.037 ± 0.241   s/op
> IcebergSourceNestedORCDataReadBenchmark.readWithProjectionFileSourceVectorized       ss    5  10.615 ± 0.336   s/op
> IcebergSourceNestedORCDataReadBenchmark.readWithProjectionIceberg                    ss    5   1.980 ± 0.167   s/op
> ```
> 
> Seems like there isn't much difference.

cc @shardulm94

File: spark/src/main/java/org/apache/iceberg/spark/data/SparkOrcValueReaders.java (hunk 3 above)
Comment (rdsr) [reply spark/src/main/java/org/apache/iceberg/spark/data/SparkOrcValueReaders.java:133]: I rerun the ORC jmh tests and do see any difference.

Comment (rdblue) [reply spark/src/main/java/org/apache/iceberg/spark/data/SparkOrcValueReaders.java:133]: Do or do not?

Comment (rdsr) [reply spark/src/main/java/org/apache/iceberg/spark/data/SparkOrcValueReaders.java:133]: do not see any differece :)

File: orc/src/main/java/org/apache/iceberg/orc/ORCSchemaUtil.java [hunk 5]
```diff
@@ -214,6 +214,7 @@ public static Schema convert(TypeDescription orcSchema) {
         "Error in ORC file, children fields and names do not match.");
 
     List<Types.NestedField> icebergFields = Lists.newArrayListWithExpectedSize(children.size());
+    // TODO how we get field ids from ORC schema
```
Comment (rdblue) [inline orc/src/main/java/org/apache/iceberg/orc/ORCSchemaUtil.java:217]: I just noticed the logic here and it's a correctness bug. ORC should not assign column IDs when one is missing. Instead, it should ignore the field.

File: orc/src/main/java/org/apache/iceberg/orc/ORCSchemaUtil.java [hunk 6]
```diff
@@ -496,7 +503,7 @@ private static boolean isRequired(TypeDescription orcType) {
     }
   }
 
-  private static int getMaxIcebergId(TypeDescription originalOrcSchema) {
+  static int getMaxIcebergId(TypeDescription originalOrcSchema) {
```
Comment (rdblue) [inline orc/src/main/java/org/apache/iceberg/orc/ORCSchemaUtil.java:506]: I don't see this used other than in the conversion logic to assign new IDs. Because that assignment is actually a correctness bug, we don't need this method at all. Also, since it isn't used anywhere else in this PR there is no need to make it package-private instead of private.

File: orc/src/main/java/org/apache/iceberg/orc/OrcSchemaWithTypeVisitor.java [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+
+      case LIST:
+        Types.ListType list = iType.asListType();
+        return visitor.array(
```
Comment (rdblue) [inline orc/src/main/java/org/apache/iceberg/orc/OrcSchemaWithTypeVisitor.java:46]: In other visitors, we try to name the method after the type in the schema that is being visited. That's why Avro uses `array` but Iceberg uses `list`. Since the category for ORC is `LIST`, should the visitor method be named `list`?

File: orc/src/main/java/org/apache/iceberg/orc/OrcValReader.java (hunk 1 above)
Comment (rdblue) [reply orc/src/main/java/org/apache/iceberg/orc/OrcValReader.java:25]: That works for me.

File: orc/src/main/java/org/apache/iceberg/orc/OrcValueReaders.java [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+    private BooleanReader() {
+    }
+
+    @Override
+    public Boolean nonNullRead(ColumnVector vector, int row) {
+      return ((LongColumnVector) vector).vector[row] != 0;
+    }
+  }
+
+  private static class ShortReader implements OrcValReader<Short> {
```
Comment (rdblue) [inline orc/src/main/java/org/apache/iceberg/orc/OrcValueReaders.java:83]: Do we need a short reader? Iceberg doesn't support short values in a table schema, so we shouldn't be expecting to ever return a short value.

File: orc/src/main/java/org/apache/iceberg/orc/OrcValueReaders.java [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+    private DoubleReader() {
+    }
+
+    @Override
+    public Double nonNullRead(ColumnVector vector, int row) {
+      return ((DoubleColumnVector) vector).vector[row];
+    }
+  }
+
+  private static class ByteReader implements OrcValReader<Byte> {
```
Comment (rdblue) [inline orc/src/main/java/org/apache/iceberg/orc/OrcValueReaders.java:143]: Similar to short, do we need this reader?

File: orc/src/main/java/org/apache/iceberg/orc/OrcValueReaders.java [hunk 10]
```diff
@@ -0,0 +1,231 @@
...
+          bytesVector.vector[row], bytesVector.start[row], bytesVector.start[row] + bytesVector.length[row]);
+    }
+  }
+
+  public abstract static class StructReader<T> implements OrcValReader<T> {
+    private final OrcValReader<?>[] readers;
+    private final int[] positions;
+    private final Object[] constants;
+
+    protected StructReader(List<OrcValReader<?>> readers) {
```
Comment (rdblue) [inline orc/src/main/java/org/apache/iceberg/orc/OrcValueReaders.java:175]: Where is this used? It doesn't look like using 0-length arrays for positions and constants is correct.

File: orc/src/main/java/org/apache/iceberg/orc/OrcValueReaders.java [hunk 11]
```diff
@@ -0,0 +1,231 @@
...
+        }
+      }
+
+      this.positions = positionList.stream().mapToInt(Integer::intValue).toArray();
+      this.constants = constantList.toArray();
+    }
+
+    protected abstract T create();
+
+    protected abstract T reuseOrCreate();
```
Comment (rdblue) [inline orc/src/main/java/org/apache/iceberg/orc/OrcValueReaders.java:201]: Why not pass the possibly reused object in here?

File: orc/src/main/java/org/apache/iceberg/orc/OrcValueReaders.java [hunk 12]
```diff
@@ -0,0 +1,231 @@
...
+      return readers[pos];
+    }
+
+    @Override
+    public T nonNullRead(ColumnVector vector, int row) {
+      StructColumnVector structVector = (StructColumnVector) vector;
+      return readInternal(create(), structVector.fields, row);
+    }
+
+    public T read(VectorizedRowBatch batch, int row) {
```
Comment (rdblue) [inline orc/src/main/java/org/apache/iceberg/orc/OrcValueReaders.java:215]: Instead of having an interface that uses `VectorizedRowBatch`, why not just have the top-level reader create a `StructColumnVector` and pass the batch's columns using that?

You can even make a reusable `StructColumnVector` to avoid needing to create a new one for every batch if you want:

```java
  class ReusableStructColumnVector extends StructColumnVector {
    public ReusableStructColumnVector(int len, ColumnVector... fields) {
      super(len, fields);
    }

    public ReusableStructColumnVector replaceVectors(ColumnVector[] fields) {
      this.fields = fields;
      return this;
    }
  }
```

Then we'd only need one `OrcValueReader` interface.

File: orc/src/main/java/org/apache/iceberg/orc/OrcValueReaders.java [hunk 13]
```diff
@@ -0,0 +1,231 @@
...
+
+  public static OrcValReader<?> floats() {
+    return FloatReader.INSTANCE;
+  }
+
+  public static OrcValReader<?> doubles() {
+    return DoubleReader.INSTANCE;
+  }
+
+  public static OrcValReader<?> bytes() {
```
Comment (rdblue) [inline orc/src/main/java/org/apache/iceberg/orc/OrcValueReaders.java:63]: These factory methods can declare the type they return, like `OrcValueReader<byte[]>`.

File: orc/src/main/java/org/apache/iceberg/orc/OrcValueReaders.java (hunk 10 above)
Comment (rdblue) [reply orc/src/main/java/org/apache/iceberg/orc/OrcValueReaders.java:175]: Nevermind, I see what's happening.

File: orc/src/main/java/org/apache/iceberg/orc/OrcValueReaders.java [hunk 14]
```diff
@@ -0,0 +1,231 @@
...
+      return readInternal(create(), structVector.fields, row);
+    }
+
+    public T read(VectorizedRowBatch batch, int row) {
+      return readInternal(reuseOrCreate(), batch.cols, row);
+    }
+
+    private T readInternal(T struct, ColumnVector[] columnVectors, int row) {
+      for (int c = 0; c < readers.length; ++c) {
+        set(struct, c, reader(c).read(columnVectors[c], row));
```
Comment (rdblue) [inline orc/src/main/java/org/apache/iceberg/orc/OrcValueReaders.java:221]: You might consider a different approach. This currently mirrors what happens in Avro, where the constants are set after reading a record. That is done because Avro can't skip fields easily -- it needs to read through a value even if the value won't be used.

But columnar formats can easily skip. That's why in Parquet, we replace the column reader with a constant reader. So the struct reader behaves exactly like normal and reads a value from every child reader. But some of those children might ignore what's in the data file and return a constant. That should be more efficient because you're not materializing columns you don't need to.

File: spark/src/main/java/org/apache/iceberg/spark/data/SparkOrcReader.java [hunk 15]
```diff
@@ -53,425 +40,83 @@
...
-  private Converter[] buildConverters() {
-    final Converter[] newConverters = new Converter[columns.size()];
-    for (int c = 0; c < newConverters.length; ++c) {
-      newConverters[c] = buildConverter(columns.get(c));
-    }
-    return newConverters;
+  @SuppressWarnings("unchecked")
+  public SparkOrcReader(
+      org.apache.iceberg.Schema expectedSchema, TypeDescription readOrcSchema, Map<Integer, ?> idToConstant) {
+    reader = (SparkOrcValueReaders.StructReader) OrcSchemaWithTypeVisitor.visit(
```
Comment (rdblue) [inline spark/src/main/java/org/apache/iceberg/spark/data/SparkOrcReader.java:52]: Minor: we prefer to use `this.reader =` when assigning to instance fields so it's clear that it is setting a field.

File: spark/src/main/java/org/apache/iceberg/spark/data/SparkOrcReader.java [hunk 16]
```diff
@@ -53,425 +40,83 @@
...
 
   @Override
   public InternalRow read(VectorizedRowBatch batch, int row) {
-    rowWriter.reset();
-    rowWriter.zeroOutNullBytes();
-    for (int c = 0; c < batch.cols.length; ++c) {
-      converters[c].convert(rowWriter, c, batch.cols[c], row);
-    }
-    return rowWriter.getRow();
+    return reader.read(batch, row);
```
Comment (rdblue) [inline spark/src/main/java/org/apache/iceberg/spark/data/SparkOrcReader.java:58]: This is where you could wrap the columns from `VectorizedRowBatch` in a `StructColumnVector`.

File: spark/src/main/java/org/apache/iceberg/spark/data/SparkOrcValueReaders.java [hunk 17]
```diff
@@ -0,0 +1,213 @@
...
+import org.apache.spark.sql.catalyst.util.MapData;
+import org.apache.spark.sql.types.Decimal;
+import org.apache.spark.unsafe.types.UTF8String;
+
+
+class SparkOrcValueReaders {
+  private SparkOrcValueReaders() {
+  }
+
+  static OrcValReader<UTF8String> strings() {
```
Comment (rdblue) [inline spark/src/main/java/org/apache/iceberg/spark/data/SparkOrcValueReaders.java:50]: Consider renaming this to something that makes it more obvious that it is returning `UTF8String`. Avro has a factory method named `utf8s` to distinguish between strings and its custom `Utf8` class.

File: spark/src/main/java/org/apache/iceberg/spark/data/SparkOrcValueReaders.java [hunk 18]
```diff
@@ -0,0 +1,213 @@
...
+    return new ArrayReader(elementReader);
+  }
+
+  static OrcValReader<?> map(OrcValReader<?> keyReader, OrcValReader<?> valueReader) {
+    return new MapReader(keyReader, valueReader);
+  }
+
+  private static class ArrayReader implements OrcValReader<ArrayData> {
+    private final OrcValReader<?> elementReader;
+    private final List<Object> reusedList = Lists.newArrayList();
```
Comment (rdblue) [inline spark/src/main/java/org/apache/iceberg/spark/data/SparkOrcValueReaders.java:73]: Container reuse requires passing objects into the read method for correctness.

The contract for container reuse is that the caller will consume the record before asking for the next one. For example, Spark might copy the entire row into an UnsafeRow.

The problem with keeping a reused list or struct in the reader itself is that the reader might be called more than once to produce a value before the row is consumed. For example, a list of structs `locations list<struct<lat:double,long:double>>` will call the inner reader for each location struct in the list before returning the record. If the reader reuses the struct, then the same struct will be added to the list multiple times and all of them will have the last values set in the reused struct.

File: spark/src/main/java/org/apache/iceberg/spark/data/SparkOrcValueReaders.java (hunk 3 above)
Comment (rdblue) [reply spark/src/main/java/org/apache/iceberg/spark/data/SparkOrcValueReaders.java:133]: The Spark/Parquet reader doesn't currently reuse containers.

File: spark/src/main/java/org/apache/iceberg/spark/data/SparkOrcValueReaders.java [hunk 19]
```diff
@@ -0,0 +1,213 @@
...
+
+      return new ArrayBasedMapData(
+          new GenericArrayData(reusedKeyList.toArray()),
+          new GenericArrayData(reusedValueList.toArray()));
+    }
+  }
+
+  static class StructReader extends OrcValueReaders.StructReader<InternalRow> {
+    private final int numFields;
+    private InternalRow internalRow;
```
Comment (rdblue) [inline spark/src/main/java/org/apache/iceberg/spark/data/SparkOrcValueReaders.java:124]: As I noted above, the reader can't keep track of a row like this.

File: spark/src/main/java/org/apache/iceberg/spark/source/RowDataReader.java [hunk 20]
```diff
@@ -109,23 +104,8 @@
...
-
-        InternalRow partition = convertToRow.apply(file.partition());
-        joined.withRight(partition);
-
-        // create joined rows and project from the joined schema to the final schema
-        iterSchema = TypeUtil.join(readSchema, partitionSchema);
-        iter = Iterators.transform(open(task, readSchema, ImmutableMap.of()), joined::withLeft);
-      }
+      iterSchema = requiredSchema;
+      iter = open(task, requiredSchema, PartitionUtil.constantsMap(task, RowDataReader::convertConstant));
```
Comment (rdblue) [inline spark/src/main/java/org/apache/iceberg/spark/source/RowDataReader.java:108]: Once this and #1004 are in, we can remove the Spark projection!

File: spark/src/test/java/org/apache/iceberg/spark/data/TestSparkOrcReader.java [hunk 21]
```diff
@@ -71,7 +71,7 @@ private void writeAndValidateRecords(Schema schema, Iterable<InternalRow> expect
 
     try (CloseableIterable<InternalRow> reader = ORC.read(Files.localInput(testFile))
         .project(schema)
-        .createReaderFunc(SparkOrcReader::new)
```
Comment (rdblue) [inline spark/src/test/java/org/apache/iceberg/spark/data/TestSparkOrcReader.java:74]: I think this should test with and without container reuse if that is implemented in this PR. Probably just make this test parameterized.

File: orc/src/main/java/org/apache/iceberg/orc/OrcValueReaders.java (hunk 12 above)
Comment (rdsr) [reply orc/src/main/java/org/apache/iceberg/orc/OrcValueReaders.java:215]: I think I did what you are suggesting before, but ran into issues. Let me try again to get the actual erorr.

**Comment by rdsr** [comment]
thanks @rdblue . Picking this up again. Should address ur comments soon

File: spark/src/main/java/org/apache/iceberg/spark/data/SparkOrcValueReaders.java (hunk 3 above)
Comment (rdsr) [reply spark/src/main/java/org/apache/iceberg/spark/data/SparkOrcValueReaders.java:133]: @rdblue . Not sure I follow  this..
> The Spark/Parquet reader doesn't currently reuse containers.

I did the JMH testing for ORC before and after my patch

File: orc/src/main/java/org/apache/iceberg/orc/ORCSchemaUtil.java (hunk 5 above)
Comment (rdsr) [reply orc/src/main/java/org/apache/iceberg/orc/ORCSchemaUtil.java:217]: Should we use another PR to fix this?

File: orc/src/main/java/org/apache/iceberg/orc/OrcValueReaders.java (hunk 8 above)
Comment (rdsr) [reply orc/src/main/java/org/apache/iceberg/orc/OrcValueReaders.java:83]: Ohk. I can change this to ints

File: spark/src/main/java/org/apache/iceberg/spark/data/SparkOrcValueReaders.java (hunk 19 above)
Comment (rdsr) [reply spark/src/main/java/org/apache/iceberg/spark/data/SparkOrcValueReaders.java:124]: In this case the reuse is only being done for the top level row and not for any internal row. Is this still a problem?

Comment (rdsr) [reply spark/src/main/java/org/apache/iceberg/spark/data/SparkOrcValueReaders.java]: I ended up removing this for now.

File: spark/src/main/java/org/apache/iceberg/spark/data/SparkOrcReader.java (hunk 16 above)
Comment (rdsr) [reply spark/src/main/java/org/apache/iceberg/spark/data/SparkOrcReader.java]: It worked!

File: orc/src/main/java/org/apache/iceberg/orc/OrcValueReaders.java (hunk 14 above)
Comment (rdsr) [reply orc/src/main/java/org/apache/iceberg/orc/OrcValueReaders.java:182]: Is is ok if I tackle this in followup?

File: orc/src/main/java/org/apache/iceberg/orc/OrcValueReaders.java (hunk 12 above)
Comment (rdsr) [reply orc/src/main/java/org/apache/iceberg/orc/OrcValueReaders.java]: It worked this time.

File: orc/src/main/java/org/apache/iceberg/orc/ORCSchemaUtil.java [hunk 22]
```diff
@@ -308,7 +309,7 @@ private static TypeDescription buildOrcProjection(Integer fieldId, Type type, bo
           orcType = convert(fieldId, type, false);
         }
     }
-
+    orcType.setAttribute(ICEBERG_ID_ATTRIBUTE, fieldId.toString());
```
[the remark this answers is not in the document]
Comment (rdsr) [reply orc/src/main/java/org/apache/iceberg/orc/ORCSchemaUtil.java:312]: Adding this is actually causing failures
```
org.apache.iceberg.data.orc.TestGenericReadProjection > testRenamedAddedField FAILED
    java.lang.IllegalArgumentException: No conversion of type LONG to self needed
        at org.apache.orc.impl.ConvertTreeReaderFactory.createAnyIntegerConvertTreeReader(ConvertTreeReaderFactory.java:1671)
        at org.apache.orc.impl.ConvertTreeReaderFactory.createConvertTreeReader(ConvertTreeReaderFactory.java:2124)
        at org.apache.orc.impl.TreeReaderFactory.createTreeReader(TreeReaderFactory.java:2331)
        at org.apache.orc.impl.TreeReaderFactory$StructTreeReader.<init>(TreeReaderFactory.java:1961)
        at org.apache.orc.impl.TreeReaderFactory.createTreeReader(TreeReaderFactory.java:2371)
        at org.apache.orc.impl.RecordReaderImpl.<init>(RecordReaderImpl.java:227)
        at org.apache.orc.impl.ReaderImpl.rows(ReaderImpl.java:752)
        at org.apache.iceberg.orc.OrcIterable.newOrcIterator(OrcIterable.java:80)
        at org.apache.iceberg.orc.OrcIterable.iterator(OrcIterable.java:65)
        at com.google.common.collect.Iterables.getOnlyElement(Iterables.java:254)
        at org.apache.iceberg.data.orc.TestGenericReadProjection.writeAndRead(TestGenericReadProjection.java:53)
```
And I vaguely remember we fixed a similar bug before in ORC

File: spark/src/test/java/org/apache/iceberg/spark/data/TestSparkOrcReader.java (hunk 21 above)
Comment (rdsr) [reply spark/src/test/java/org/apache/iceberg/spark/data/TestSparkOrcReader.java:74]: For now I've removed the reuse code. We can tackle than in followup

**Comment by rdsr** [comment]
@rdblue This is ready for another round of review

File: orc/src/main/java/org/apache/iceberg/orc/ORCSchemaUtil.java (hunk 22 above)
[the remark this answers is not in the document]
Comment (rdblue) [reply orc/src/main/java/org/apache/iceberg/orc/ORCSchemaUtil.java:311]: It would be great to know what's going on here. Since this is just a projection schema and the reader is built with the Iceberg schema (that has required/optional), I don't think it is really a blocker. But setting a property here shouldn't cause ORC to fail, right?

File: orc/src/main/java/org/apache/iceberg/orc/OrcValueReaders.java (hunk 14 above)
Comment (rdblue) [reply orc/src/main/java/org/apache/iceberg/orc/OrcValueReaders.java:182]: Yeah, that sounds good.

File: spark/src/main/java/org/apache/iceberg/spark/source/RowDataReader.java (hunk 20 above)
Comment (rdblue) [reply spark/src/main/java/org/apache/iceberg/spark/source/RowDataReader.java:108]: #1004 is in. That might be why this has conflicts.

**Comment by rdsr** [comment]
Weird. I thought I did resolve the conflicts.

File: orc/src/main/java/org/apache/iceberg/orc/ORCSchemaUtil.java (hunk 22 above)
[the remark this answers is not in the document]
Comment (rdsr) [reply orc/src/main/java/org/apache/iceberg/orc/ORCSchemaUtil.java:311]: I'll file the necessary followups.

**Comment by rdsr** [comment]
@rdblue . Fixed conflicts and filed all the necessary followups.

**Comment by rdblue** [comment]
Looks good to me! I'll merge it. Thanks, @rdsr!
